core/src/main/java/lucee/runtime/db/QoQ.java

```diff
@@ -73,205 +96,423 @@ public Query execute(PageContext pc, SQL sql, int maxrows) throws PageException
+			// table" than bail after a single row
+			if (hasAggregateSelect) {
+				break;
+			}
+		}
+
+	}
+
+	/**
+	 * Process a single select that is partitioned (grouped)
+	 * 
+	 * @param pc PageContext
+	 * @param select Select instance
+	 * @param source Query we're selecting from
+	 * @param target Query object we're adding rows into. (passed back by reference)
+	 * @param maxrows
+	 * @param sql
+	 * @param hasOrders Is this overall query ordered?
+	 * @param isUnion Is this part of a union?
+	 * @param trgColumns Lookup array of column
+	 * @param trgValues Lookup array of expressions
+	 * @param headers select columns
+	 * @throws PageException
+	 */
+	private void executeSinglePartitioned(PageContext pc, Select select, Query source, QueryImpl target, int maxrows, SQL sql, boolean hasOrders, boolean isUnion,
+			QueryColumn[] trgColumns, Object[] trgValues, Collection.Key[] headers) throws PageException {
+
+		Operation where = select.getWhere();
+		// Initialize object to track our partioned data
+		QueryPartitions queryPartitions = new QueryPartitions(sql, select.getSelects(), select.getGroupbys(), target, select.getAdditionalColumns(), this);
+
+		// For all records in the source query
+		for (int row = 1; row <= source.getRecordcount(); row++) {
+			sql.setPosition(0);
+			// If where operation is matched (or doesn't exist) ....
+			if (where == null || Caster.toBooleanValue(executeExp(pc, sql, source, where, row))) {
+				// ... add this row to our partitioned data
+				queryPartitions.addRow(pc, source, row, false);
+			}
+		}
+
+		// Now that all rows are partioned, eliminate partions we don't need via the having clause
+		if (select.getHaving() != null) {
+			// Loop over each partition
+			for (String partionKey: queryPartitions.getPartitions().keySet().toArray(new String[queryPartitions.getPartitions().size()])) {
+				// Eval the having clause on it
+				if (!Caster.toBooleanValue(executeExp(pc, sql, queryPartitions.getPartitions().get(partionKey), select.getHaving(), 1))) {
+					// Voted off the island :/
+					queryPartitions.getPartitions().remove(partionKey);
 				}
+				// ColumnExpressions cache the actual query column they use internally
+				// need to reset any cache data in the having Expression
+				// since each iteration is on a diff Query instance
+				select.getHaving().reset();
 			}
 		}
 
-		// Group By
-		if (select.getGroupbys().length > 0) throw new DatabaseException("group by are not supported at the moment", null, sql, null);
-		if (select.getHaving() != null) throw new DatabaseException("having is not supported at the moment", null, sql, null);
+		// Add first row of each group of partitioned data into final result
+		QueryImpl sourcePartition;
+		for (String partionKey: queryPartitions.getPartitions().keySet().toArray(new String[queryPartitions.getPartitions().size()])) {
+			sourcePartition = queryPartitions.getPartitions().get(partionKey);
```

Comment:
again 
`for(QueryImpl sourcePartition:...values())`

Reply:
Changed.